src/main/scala/mesosphere/marathon/core/launcher/impl/InstanceOpFactoryImpl.scala

```diff
@@ -219,13 +217,12 @@ class InstanceOpFactoryImpl(
           ResourceSelector.reservable, config, schedulerPlugins, localRegion)
       resourceMatchResponse match {
         case matches: ResourceMatchResponse.Match =>
-          val instance = scheduledInstances.find(!_.hasReservation).getOrElse(throw new IllegalStateException(s"Expecting to have scheduled instance without reservation but non is found in: $scheduledInstances"))
```

Comment:
Nice cleanup. However, we lost this exception here. Now, in case of a bug, nothing will happen and we'll wonder why that's the case. I'd very much like to have it back.

Reply:
This bug cannot happen anymore. This exception would only throw if `needToReserve` was true but `scheduledInstances.find(!_.hasReservation)` was empty. However, `val needToReserve = scheduledInstances.exists(!_.hasReservation)`. So this exception could not happen unless `scheduledInstances` was changed.

The compiler checks the semantics for us in `scheduledInstances.find(!_.hasReservation).map { firstScheduledInstance ... }`.